draft-ietf-quic-recovery.md

```diff
@@ -398,8 +417,80 @@ A packet sent on an RTO alarm MUST NOT be blocked by the sender's congestion
+losing the client's Initial, but receiving the 0-RTT packets.  Below are
+three different optimizations in increasing complexity that minimize
+handshake latency.
+
+### Empty ACK frames
+
+The EMPTY_ACK frame enables faster recovery of lost Initial and Handshake
+packets when the only other outstanding packets are undecryptable.
+In that case, the peer can’t decrypt the packet or packet number,
+but they do know a packet for a QUIC connection of a supported version was
+received.
+
+The receiver SHOULD send an EMPTY_ACK frame soon (e.g., 1ms) after
+undecryptable packets are received, even if those received packets are
+not buffered for later decryption.  The small delay allows for cases when
+0-RTT packets are reordered in front of the Initial, which is not uncommon
+on networks that prioritize small packets.  The receiver should limit the
+number of EMPTY_ACK frames sent to one per packet number space per RTT.
+
+When an EMPTY_ACK frame is received, a sender should immediately
+retransmit the missing handshake packet(s) as though the handshake timer
+fired and re-arm the handshake timer when the handshake packets are sent.
+If the missing handshake data was timer retransmitted after the packets
+that triggered the empty ack, then the empty ack should be ignored.
+
+An EMPTY_ACK frame does not acknowledge a new packet, and in cases
+when multiple packets are outstanding, the RTT signal is ambiguous,
+so it should not be used like an RTT signal from a newly acknowledged
+packet.  It MAY change the connection’s default RTT if no RTT measurements
+have been taken.
+
+### Coalesced Packets
+
+Despite loss recovery being separate for different packet number spaces,
+the ability to send a coalesced packet enables faster recovery with small,
+and sometimes no overhead.  The acknowledgement of a coalesced packet
+allows QUIC recovery to use early retransmit to determine if any prior
+packets in that space were lost without waiting for timeouts.
+
+This optimization is particularly useful when:
+
+ * Retransmitting the client’s Initial, which must be padded to a full
+   sized packet, so the datagram typically has extra space to retransmit
+   some outstanding 0-RTT data.
+ * The server sends 1-RTT data soon after the final handshake flight
+   (containing the server Finished) and can proactively retransmit an empty
+   CRYPTO_HS frame bundled with one or more 1-RTT packets.
+ * The clients sends 1-RTT data soon after the final TLS flight
+   (containing the client Finished) and can proactively retransmit the
+   final client flight with one or more 1-RTT packets.
+
+### Implicit Acknowledgements
+
+Handshake data may be cancelled by handshake state transitions.
+
+In particular:
+
+ * A peer processing data in a HANDSHAKE packet indicates
+   the INITIAL packet(s) have been delivered.
+ * A peer processing 1-RTT packets indicates all CRYPTO_HS data in
+   HANDSHAKE packets has been delivered.
```

Comment:
does this interact OK with CFIN and 0-RTT data given that there is packet number space for both 0-RTT and 1-RTT data?

I presume what this provision means, in the case of 0-rtt, if a client receives an ACK for something it sent under 1-RTT keys then it can assume the server recevied CFIN and therefore the Handshake Keys are no longer needed..

but 0-RTT and 1-RTT packets share the same packet number space.. and you'll definitely have 1-RTT encoded packets from the server that contain acks of 0-RTT data (in the same packet number space of course) that are sent before the server has received CFIN.

So this provision requires the client to figure out the key used to encrypt the data that is being acked to do the right thing. That seems to go against the general grain of the whole design here.

Reply:
I updated the text and I noted that processing 0RTT means nothing.

But yeah, you're right that when the client processes 0.5RTT data, it doesn't mean the client cert or CFIN have been received.

Any suggestions on describing this more correctly?

Reply:
I think the situation is a little more nuanced.  For a client to generate 1-RTT packets at all, it has to have received the server's handshake packets in full.  That means a server can consider all of its packets acknowledged when it receives a correct 1-RTT packet from the client, though that can't happen until it has also received the end of the client's handshake.

From the client's side, it knows the end of its handshake has been received when the server ACKs a 1-RTT packet.  So it has to remember the packet number of the first packet it sent under 1-RTT keys, and an ACK for anything larger than that acknowledges all the handshake data.

They both hinge on the client 1-RTT packets.  When the server acknowledges a 1-RTT packet, it marks all its handshake data received.  When the client receives that same acknowledgement, it marks all of *its* handshake data received.